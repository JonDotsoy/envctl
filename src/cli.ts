import {
  flags,
  command,
  restArgumentsAt,
  rule,
  type Rule,
  makeHelmMessage,
  flag,
  isBooleanAt,
} from "@jondotsoy/flags";
import * as subcommands from "./subcommands";

const main = async (args: string[]) => {
  type Options = {
    build: string[];
    list: string[];
    init: string[];
    showHelp: boolean;
  };

  const rules: Rule<Options>[] = [
    rule(flag("--help", "-h"), isBooleanAt("showHelp"), {
      description: "Show help to envctl",
    }),
    rule(command("use"), restArgumentsAt("build"), {
      description: "Build the env file. Example: envctl use <context>",
    }),
    rule(command("list"), restArgumentsAt("list"), {
      description: "List environment available",
    }),
    rule(command("init"), restArgumentsAt("init"), {
      description: "Init the template file",
    }),
  ];

  const makeHelp = () =>
    makeHelmMessage("envctl", rules, ["build --ctx=<ctx>"]);

  const options = flags<Options>(args, {}, rules);

  if (options.showHelp) return console.error(makeHelp());

  for (const [subcommand, handler] of Object.entries(subcommands)) {
    if (Reflect.has(options, subcommand)) {
      return handler(Reflect.get(options, subcommand));
    }
  }

  return console.error(makeHelp());
};

await main(process.argv.slice(2)).catch((ex) => {
  if (ex instanceof Error) {
    return console.error(ex.message);
  }
  throw ex;
});
