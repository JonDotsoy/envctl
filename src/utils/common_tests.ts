import fs from "fs/promises";
import process from "node:process";
import { useWorkspace as createUseWorkspace } from "use-workspace";

namespace WorkspacesName {
  const toCharCode = (value: string) => value.charCodeAt(0);
  const range = (start: number, end: number) =>
    new Array(Math.max(0, end + 1 - start))
      .fill(null)
      .map((_, index) => start + index);

  const charCodeSerializable = [
    toCharCode("_"),
    toCharCode("-"),
    ...new Uint8Array(range(toCharCode("a"), toCharCode("z"))),
    ...new Uint8Array(range(toCharCode("A"), toCharCode("Z"))),
    ...new Uint8Array(range(toCharCode("0"), toCharCode("9"))),
  ];

  const isSerializable = (charCode: number) =>
    charCodeSerializable.includes(charCode);

  export const serialize = (value: string) => {
    return new TextDecoder().decode(
      new Uint8Array(
        Array.from(new TextEncoder("utf-8").encode(value)).flatMap(
          (charCode) =>
            isSerializable(charCode)
              ? charCode
              : [
                  toCharCode("["),
                  ...new TextEncoder().encode(`${charCode}`),
                  toCharCode("]"),
                ],
        ),
      ),
    );
  };

  export const deserialize = (value: string) => {
    let indexAtSkip = -1;
    return new TextDecoder().decode(
      new Uint8Array(
        Array.from(new TextEncoder("utf-8").encode(value)).flatMap(
          (charCode, index, charCodes) => {
            if (charCode === toCharCode("[")) {
              const indexClosed = charCodes.indexOf(toCharCode("]"), index);
              if (indexClosed) {
                const chunk = charCodes.slice(index + 1, indexClosed);
                const charCode = Number(
                  new TextDecoder().decode(new Uint8Array(chunk)),
                );
                indexAtSkip = indexClosed;
                return charCode;
              }
            }

            if (index <= indexAtSkip) return [];

            return charCode;
          },
        ),
      ),
    );
  };
}

const workspacesLocation = new URL(`__workspaces__/`, import.meta.url);
const getWorkspaceLocation = (workspacesName: string) =>
  new URL(`${WorkspacesName.serialize(workspacesName)}/`, workspacesLocation);

export const useWorkspace = async (name: string = "default") => {
  const { location: workspaceLocation } = await createUseWorkspace(name);

  const toLocation = (relativeName: string) =>
    new URL(relativeName, workspaceLocation);

  const writeFile = async (
    relativeName: string,
    payload: Uint8Array,
    options?: { mode?: number },
  ) => {
    await fs.mkdir(new URL("./", toLocation(relativeName)), {
      recursive: true,
    });
    await fs.writeFile(toLocation(relativeName), payload, {
      mode: options?.mode,
    });
  };

  const readFile = async (relativeName: string) => {
    return fs.readFile(toLocation(relativeName));
  };

  await writeFile(".gitignore", new TextEncoder().encode("*"));

  return {
    workspaceLocation,
    toLocation,
    writeFile,
    readFile,
  };
};

/**
 *
 * @example
 * using useChdir()
 */
export const useChdir = (directory: URL) => {
  const cwd = () => process.cwd();

  const initialCWD = process.cwd();
  process.chdir(directory.pathname);

  const restore = () => {
    process.chdir(initialCWD);
  };

  return {
    cwd,
    restore,
    initial: {
      cwd: initialCWD,
    },
    [Symbol.dispose]() {
      restore();
    },
  };
};
