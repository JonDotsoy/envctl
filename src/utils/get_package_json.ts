import pkg from "../../package.json" with { type: "json" };

export const getPackageJson = () => pkg;
