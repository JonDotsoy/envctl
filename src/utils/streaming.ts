export type MessageText = string;
export type MessageObject = {
  type: "INFO" | "LOG" | "ERROR";
  value: MessageText;
};
export type Message = MessageText | MessageObject;
export type Streaming = ReadableStream<Message>;
export const isMessageText = (value: Message): value is MessageText =>
  typeof value === "string";
export const isMessageObject = (value: Message): value is MessageObject =>
  typeof value === "object";

export const messageToText = (message: Message): MessageText =>
  isMessageText(message) ? message : message.value;
export const messageToObject = (message: Message): MessageObject =>
  isMessageObject(message) ? message : { type: "LOG", value: message };

export const streamingToArray = async (streaming: Streaming) => {
  return (await Array.fromAsync(streaming, messageToObject)) as MessageObject[];
};

class GlobalMessages {
  private readablePromise =
    Promise.withResolvers<Bun.ReadableStreamController<Message>>();

  private subs = new Set<(message: Message) => any>();

  readonly writableSubs = new WritableStream<Message>({
    write: (message) => {
      for (const sub of this.subs) {
        sub(message);
      }
    },
  });

  readonly readable = new ReadableStream<Message>({
    start: (controller) => {
      this.readablePromise.resolve(controller);
    },
  }).pipeTo(this.writableSubs);

  async getController() {
    return this.readablePromise.promise;
  }

  async close() {
    const controller = await this.getController();
    controller.close();
  }

  subscribe(cb: (message: Message) => any) {
    this.subs.add(cb);
    return () => {
      this.subs.delete(cb);
    };
  }
}

export const globalMessages = new GlobalMessages();

export const useMessages = async () => {
  const controller = await globalMessages.getController();

  const send = (message: Message) =>
    controller.enqueue(messageToObject(message));

  return {
    send,
  };
};

export const catchToMessages = async (cb: () => Promise<void>) => {
  const messages = await useMessages();

  try {
    return await cb();
  } catch (ex) {
    if (ex instanceof Error)
      return messages.send({ type: "ERROR", value: `${ex.message}\n` });
    throw ex;
  }
};
