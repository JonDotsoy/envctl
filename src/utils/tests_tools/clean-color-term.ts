export const cleanColorTerm = (str: string) => {
  let closedUntil: number = -1;
  return new TextDecoder().decode(
    new Uint8Array(
      Array.from(new TextEncoder().encode(str)).flatMap(
        (charCode, index, a) => {
          if (index < closedUntil) return [];
          if (charCode === 0x1b) {
            const closingSymbol = a.indexOf(0x6d, index);
            closedUntil = closingSymbol + 1;
            return [];
          }
          return charCode;
        },
      ),
    ),
  );
};
