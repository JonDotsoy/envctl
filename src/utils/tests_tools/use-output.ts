import { readableStreamToText } from "bun";

export const useOutput = async () => {
  const outputControllerPromise =
    Promise.withResolvers<Bun.ReadableStreamController<string>>();
  const outputStream = new ReadableStream<string>({
    start: (controller) => outputControllerPromise.resolve(controller),
  });
  const outputController = await outputControllerPromise.promise;

  let alreadyClosed = false;
  const close = () => {
    if (!alreadyClosed) {
      outputController.close();
      alreadyClosed = true;
    }
  };

  return {
    send: (message: string) => outputController.enqueue(message),
    close: close,
    subscribeReadable: (readable: ReadableStream<string>) =>
      readable.pipeTo(
        new WritableStream<string>({
          write: (message) => outputController.enqueue(message),
        }),
      ),
    stream: outputStream,
    controller: outputController,
    toText: () => readableStreamToText(outputStream),
    [Symbol.dispose]: () => {
      close();
    },
  };
};
