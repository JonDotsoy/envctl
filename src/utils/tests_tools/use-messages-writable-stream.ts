import { messageToText, type Message, globalMessages } from "../streaming";
import { cleanColorTerm } from "./clean-color-term";

export const useMessagesWritableStream = () => {
  let output = ``;
  const unSubs = new Set<() => void>();

  const cb = (message: Message) => {
    output = `${output}${cleanColorTerm(messageToText(message))}`;
  };

  return {
    toText() {
      return output;
    },
    connect: (obj: {
      subscribe: (cb: (message: Message) => void) => () => any;
    }) => {
      unSubs.add(obj.subscribe(cb));
    },
    [Symbol.dispose]() {
      for (const unSub of unSubs) {
        unSub();
      }
    },
  };
};

export const useGlobalMessages = () => {
  const messagesWritableStream = useMessagesWritableStream();

  messagesWritableStream.connect(globalMessages);

  return {
    toText: () => messagesWritableStream.toText(),
    [Symbol.dispose]: () => {
      messagesWritableStream[Symbol.dispose]();
    },
  };
};
