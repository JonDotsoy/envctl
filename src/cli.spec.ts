import { test, expect } from "bun:test";
import { useChdir, runSample, useWorkspace } from "./utils/common_tests";
import { cli } from "./cli";
import { catchToMessages } from "./utils/streaming";
import { useGlobalMessages } from "./utils/tests_tools/use-messages-writable-stream";
import fs from "fs/promises";

test("run cli", async () => {
  const { workspaceLocation, writeFile } = await useWorkspace();
  using _chdir = useChdir(workspaceLocation);
  using messagesWritableStream = useGlobalMessages();

  await writeFile(".env.local", new TextEncoder().encode("FOO=biz"));
  await writeFile(".env.template", new TextEncoder().encode("FOO=taz"));

  await catchToMessages(() => cli([]));

  expect(messagesWritableStream.toText()).toMatchSnapshot();
});

test("run use local env (using full path)", async () => {
  const { workspaceLocation, writeFile, readFile } = await useWorkspace();
  using _chdir = useChdir(workspaceLocation);
  using messagesWritableStream = useGlobalMessages();

  await writeFile(".env.local", new TextEncoder().encode("FOO=biz"));
  await writeFile(".env.template", new TextEncoder().encode("FOO=taz"));

  await catchToMessages(() => cli(["use", ".env.local"]));

  expect(messagesWritableStream.toText()).toMatchSnapshot();
  expect(new TextDecoder().decode(await readFile(".env"))).toMatchSnapshot();
});

test("run use local env (using partial path)", async () => {
  const { workspaceLocation, writeFile, readFile } = await useWorkspace();
  using _chdir = useChdir(workspaceLocation);
  using messagesWritableStream = useGlobalMessages();

  await writeFile(".env.local", new TextEncoder().encode("FOO=biz"));
  await writeFile(".env.template", new TextEncoder().encode("FOO=taz"));

  await catchToMessages(() => cli(["use", "local"]));

  expect(messagesWritableStream.toText()).toMatchSnapshot();
  expect(new TextDecoder().decode(await readFile(".env"))).toMatchSnapshot();
});

test("run use local env (using subpath location)", async () => {
  const { workspaceLocation, writeFile, readFile } = await useWorkspace();
  using _chdir = useChdir(workspaceLocation);
  using messagesWritableStream = useGlobalMessages();

  await writeFile(".envs/local.env", new TextEncoder().encode("FOO=biz"));
  await writeFile(".env.template", new TextEncoder().encode("FOO=taz"));

  await catchToMessages(() => cli(["use", ".envs/local.env"]));

  expect(messagesWritableStream.toText()).toMatchSnapshot();
  expect(new TextDecoder().decode(await readFile(".env"))).toMatchSnapshot();
});

test("run use local env (using subpath relative location)", async () => {
  const { workspaceLocation, writeFile, readFile } = await useWorkspace();
  using _chdir = useChdir(workspaceLocation);
  using messagesWritableStream = useGlobalMessages();

  await writeFile(".envs/local.env", new TextEncoder().encode("FOO=biz"));
  await writeFile(".env.template", new TextEncoder().encode("FOO=taz"));

  await catchToMessages(() => cli(["use", "local"]));

  expect(messagesWritableStream.toText()).toMatchSnapshot();
  expect(new TextDecoder().decode(await readFile(".env"))).toMatchSnapshot();
});

test("should call to envctl list and display all envs files found", async () => {
  const workspace = await useWorkspace();
  using _chdir = useChdir(workspace.workspaceLocation);
  using messages = useGlobalMessages();

  await workspace.writeFile(".env.template", new TextEncoder().encode(``));
  await workspace.writeFile(".envs/template", new TextEncoder().encode(``));

  await workspace.writeFile(".env.aaa", new TextEncoder().encode(``));
  await workspace.writeFile(".bbb.env", new TextEncoder().encode(``));
  await workspace.writeFile(".envs/ccc", new TextEncoder().encode(``));
  await workspace.writeFile(".envs/.ddd", new TextEncoder().encode(``));
  await workspace.writeFile(".envs/eee.env", new TextEncoder().encode(``));
  await workspace.writeFile(".envs/.fff.env", new TextEncoder().encode(``));

  await catchToMessages(() => cli(["list"]));

  await new Promise((r) => setTimeout(r, 100));
  expect(messages.toText()).toMatchSnapshot();
});

test("should call to envctl list and display json output", async () => {
  const workspace = await useWorkspace();
  using _chdir = useChdir(workspace.workspaceLocation);
  using messages = useGlobalMessages();

  await workspace.writeFile(".env.template", new TextEncoder().encode(``));
  await workspace.writeFile(".envs/template", new TextEncoder().encode(``));

  await catchToMessages(() => cli(["list", "-j"]));

  await new Promise((r) => setTimeout(r, 100));
  expect(messages.toMessages()).toMatchSnapshot();
});

test("should call to envctl version to display version number", async () => {
  const workspace = await useWorkspace();
  using _chdir = useChdir(workspace.workspaceLocation);
  using messages = useGlobalMessages();

  await catchToMessages(() => cli(["version"]));

  await new Promise((r) => setTimeout(r, 100));
  expect(JSON.stringify(messages.toMessages())).toMatch(
    /Version v\d+\.\d+\.\d+/,
  );
});

test("should call to envctl version to display version number (json output)", async () => {
  const workspace = await useWorkspace();
  using _chdir = useChdir(workspace.workspaceLocation);
  using messages = useGlobalMessages();

  await catchToMessages(() => cli(["version", "--json"]));

  await new Promise((r) => setTimeout(r, 100));
  expect(JSON.stringify(messages.toMessages())).toMatch(`\\"version\\":`);
  expect(JSON.stringify(messages.toMessages())).toMatch(/\d+\.\d+\.\d+/);
});

test("should call to envctl use without template", async () => {
  const workspace = await useWorkspace();
  using _chdir = useChdir(workspace.workspaceLocation);
  using messages = useGlobalMessages();

  await workspace.writeFile(".envs/staging", new TextEncoder().encode("foo"));

  await catchToMessages(() => cli(["use", "staging"]));

  await new Promise((r) => setTimeout(r, 100));
  console.log("ðŸš€ ~ test.only ~ messages.toMessages():", messages.toMessages());

  // expect(JSON.stringify(messages.toMessages())).toMatch(`\\"version\\":`);
  // expect(JSON.stringify(messages.toMessages())).toMatch(/\d+\.\d+\.\d+/);
});

test("should create a .env with espacial characters values", async () => {
  const { workspaceLocation, toLocation, writeFile, readFile } =
    await useWorkspace();
  using _chdir = useChdir(workspaceLocation);
  using messages = useGlobalMessages();

  await writeFile(
    ".envs/staging",
    new TextEncoder().encode(`FOO="biz taz"\nBIZ="aaa\nbbb"`),
  );
  await writeFile(
    ".env.example",
    new TextEncoder().encode(`FOO= # comment\nBIZ=`),
  );

  await catchToMessages(() => cli(["use", "staging"]));
  await new Promise((r) => setTimeout(r, 100));

  expect(messages.toMessages()).toMatchSnapshot("messages");
  expect(new TextDecoder().decode(await readFile(".env"))).toMatchSnapshot(
    ".env payload",
  );

  await fs.copyFile(
    toLocation(".env"),
    new URL("./samples/.env", import.meta.url),
  );
  await runSample(new URL("./samples/sample1.ts", import.meta.url));
});
