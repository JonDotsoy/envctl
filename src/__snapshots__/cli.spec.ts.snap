// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`run cli 1`] = `
"Usage: envctl use <ctx>

flag:
    --help,      Show help to envctl
    -h

command:
    use       Build the env file. Example: envctl use <context>
    list      List environment available
    version   Show version enabled
    init      Init the template file

"
`;

exports[`run use local env (using full path) 1`] = `
"Created env by .env.local file
"
`;

exports[`run use local env (using full path) 2`] = `"FOO="biz""`;

exports[`run use local env (using partial path) 1`] = `
"Created env by .env.local file
"
`;

exports[`run use local env (using partial path) 2`] = `"FOO="biz""`;

exports[`run use local env (using subpath location) 1`] = `
"Created env by .envs/local.env file
"
`;

exports[`run use local env (using subpath location) 2`] = `"FOO="biz""`;

exports[`run use local env (using subpath relative location) 1`] = `
"Created env by .envs/local.env file
"
`;

exports[`run use local env (using subpath relative location) 2`] = `"FOO="biz""`;

exports[`should call to envctl list and display all envs files found 1`] = `
"- aaa (.env.aaa)
- bbb (.bbb.env)
- ddd (.envs/.ddd)
- cc (.envs/ccc)
- fff (.envs/.fff.env)
- eee (.envs/eee.env)
"
`;

exports[`should call to envctl list and display json output 1`] = `
[
  {
    "type": "ERROR",
    "value": 
"No contexts found. You can create one by making a .env.<ctx> file and adding desired variables.
"
,
  },
  {
    "type": "LOG",
    "value": "[]",
  },
]
`;

exports[`should create a .env with espacial characters values: messages 1`] = `
[
  {
    "type": "INFO",
    "value": 
"Created env by .envs/staging file
"
,
  },
]
`;

exports[`should create a .env with espacial characters values: .env payload 1`] = `
"FOO="biz taz"
"
`;
